docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -453,70 +471,62 @@ spec:
 
-##### Cluster upgrade (using create-swap-and-delete)
+##### KubeadmControlPlane rollout (using create-swap-and-delete)
 
-- Triggered on any changes to Version or the InfrastructureTemplate
-  - The KubeadmConfiguration, for the first iteration, is treated and validated as immutable and changes to it do not trigger an upgrade
-  - Must be able to trigger an upgrade for the underlying machine image (e.g. nothing in the above list has changed but an upgrade is required)
-      - By adding a timestamp field, `upgradeAfter`, which can be set to a specific time in the future
+- Triggered by: 
+    - Changes to Versio
+    - Changes to the kubeadmConfigSpec
+    - Changes to the infrastructureRef
+    - The `upgradeAfter` field, which can be set to a specific time in the future
         - Set to `nil` or the zero value of `time.Time` if no upgrades are desired
         - An upgrade will run when that timestamp is passed
         - Good for scheduling upgrades/SLOs
         - Set `upgradeAfter` to now (in RFC3339 form) if an upgrade is required immediately
 
-- Upgrade operations rely on scale up and scale down which are be blocked based on Etcd and control plane health checks
+- Rollout operations rely on scale up and scale down which are be blocked based on Etcd and control plane health checks
   - See [Health checks](#Health checks) below.
 
-- The upgrade algorithm is as follows:
-  - Find Machines that need upgrading
-    - Configuration hash doesn't match
-    - `upgradeAfter` field is before now
-  - Mark machine as selected for upgrade
-  - Scale up control plane
-  - Mark machine as having a replacement
-  - Scale down control plane by removing a machine that has been marked for removal and has a replacement already created
-
-- Determining if a Machine is "up to date" will be done through the use of a label (controlplane.cluster.x-k8s.io/hash) that is placed on the Machine at creation time. The value of this label is generated by computing the Hash of the KubeadmControlPlaneSpec (minus the replicas field). This would allow triggering an Upgrade based on any changes to Version, InfrastructureTemplate, or the KubeadmConfiguration used (when a new configuration is referenced in a new InfrastructureTemplate).
-  - For example, a CVE is patched in containerd and updated images are available. Note that the Kubernetes version is not changed. To trigger an upgrade, the user updates the image in the InfrastructureTemplate (as in the Acme cluster example above, the image is stored in InfrastructureTemplate.Spec.OSImage.ID).
+- The rollout algorithm is the following:
+  - Find Machines that need rollout
+  - If there is a machine requiring rollout
+    - Scale up control plane creating a machine with the new spec
+    - Scale down control plane by removing one of the machine that needs rollout (the oldest out-of date machine in the failure domain that has the most out-of-date control-plane machines on it).
```

Comment:
```suggestion
    - Scale down control plane by removing one of the machine that needs rollout (the oldest out-of date machine in the failure domain that has the most control-plane machines on it).
```

Reply:
It is picking a failure domain that has the most number of control plane machines and at least 1 outdated machine in it.

Reply:
If I'm not wrong this is considering only out of date machines:

In `upgradeControlPlane`, the we are calling `scaleDownControlPlane` and passing the list of `machinesRequireUpgrade`
https://github.com/kubernetes-sigs/cluster-api/blob/38d8699bb7c42f10b20a36a553161ba85f739fc7/controlplane/kubeadm/controllers/upgrade.go#L98

Such list is the used for `selectMachineForScaleDown`
https://github.com/kubernetes-sigs/cluster-api/blob/494b8b146055390d0fa49910f5ebdf02e83199d5/controlplane/kubeadm/controllers/scale.go#L103

And being the list not empty it drives also the selection of the failure domains to start deleting from

https://github.com/kubernetes-sigs/cluster-api/blob/494b8b146055390d0fa49910f5ebdf02e83199d5/controlplane/kubeadm/controllers/scale.go#L152-L155

Reply:
While deciding on the failure domain, we use all control-plane machine (c.machines):
https://github.com/kubernetes-sigs/cluster-api/blob/494b8b146055390d0fa49910f5ebdf02e83199d5/controlplane/kubeadm/internal/failure_domain.go#L54-L56

Reply:
FailureDomainWithMostMachines is actually considering all machines not just `out-of-date` machines. It picks an fd with most machines that has at least 1 out-of-date machine.

Reply:
done!